src/IO/ReadWriteBufferFromHTTP.h

```diff
@@ -222,41 +297,110 @@ namespace detail
         }
 
         bool nextImpl() override
         {
             if (next_callback)
                 next_callback(count());
 
-            if (use_external_buffer)
+            if (read_range.end && bytes_read == read_range.end.value())
+                return false;
+
+            if (impl)
             {
-                /**
-                * use_external_buffer -- means we read into the buffer which
-                * was passed to us from somewhere else. We do not check whether
-                * previously returned buffer was read or not (no hasPendingData() check is needed),
-                * because this branch means we are prefetching data,
-                * each nextImpl() call we can fill a different buffer.
-                */
-                impl->set(internal_buffer.begin(), internal_buffer.size());
-                assert(working_buffer.begin() != nullptr);
-                assert(!internal_buffer.empty());
+                if (use_external_buffer)
+                {
+                    /**
+                    * use_external_buffer -- means we read into the buffer which
+                    * was passed to us from somewhere else. We do not check whether
+                    * previously returned buffer was read or not (no hasPendingData() check is needed),
+                    * because this branch means we are prefetching data,
+                    * each nextImpl() call we can fill a different buffer.
+                    */
+                    impl->set(internal_buffer.begin(), internal_buffer.size());
+                    assert(working_buffer.begin() != nullptr);
+                    assert(!internal_buffer.empty());
+                }
+                else
+                {
+                    /**
+                    * impl was initialized before, pass position() to it to make
+                    * sure there is no pending data which was not read.
+                    */
+                    if (!working_buffer.empty())
+                        impl->position() = position();
+                }
             }
-            else
+
+            bool result = false;
+            size_t milliseconds_to_wait = settings.http_retry_initial_backoff_ms;
+
+            for (size_t i = 0; i < settings.http_max_tries; ++i)
             {
-                /**
-                * impl was initialized before, pass position() to it to make
-                * sure there is no pending data which was not read.
-                */
-                if (!working_buffer.empty())
-                    impl->position() = position();
+                try
+                {
+                    if (!impl)
+                    {
+                        /// If error is not retriable -- false is returned and exception is set.
+                        /// Otherwise the error is thrown and retries continue.
+                        bool initialized = initialize();
+                        if (!initialized && exception)
```

Comment:
Is it possible that `initialize()` returned false, but `exception` is `nullptr`? Or, for example, that `initialize` returned `true`, but `impl` is `nullptr`? Let's either remove unnecessary flags or add assertions.